docs/proposal/consistent_hash.md

```diff
@@ -0,0 +1,171 @@
+---
+title: Consistent hash LB
+authors:
+- "@bfforever" # Authors' GitHub accounts here.
+reviewers:
+- "@supercharge-xsy"
+- "@hzxuzhonghu"
+- "@nlwcy"
+- TBD
+approvers:
+- "@robot"
+- TBD
+
+creation-date: 2024-05-29
+
+---
+
+## Your short, descriptive title
+
+Add Consistent hash based LB in ebpf prog
+
+### Summary
+
+<ol>
+<li>Maglev consistent hash algorithm</li>
+<li>IP tuple info based hash</li>
+<li>L7 http header based hash</li>
+</ol>
+
+#### Goals
+
+<ol>
+<li>Networking load balance</li>
+<li>Guarrant a client conn always send request to same backend</li>
+<li>Minimize remapping</li>
+</ol>
+
+### Proposal
+
+#### maglev
+Algorithm main purpose: map all backends into a fixed look up table, require even mapping, and try to minimize mapping changes when backend change (remove or add a backend), avoid global remapping.
+
+
+#### ebpf implement situation or point
+<ol>
+<li>Cgroup level</li>
+In cgroup socket implement l4 and l7 header based consistent hash.
+<li>Container veth tc level</li>
+In tc level implement l4 based and l7 based consistent hash.
+</ol>
+
+
+### Design Details
+
+![consistent_hash](pics/kmesh_consistent_hash.svg)
+#### maglev 
+main steps:
+
+<ol>
+<li>Map endpoints of Cluster to a fixed table</li>
+Based on Cluster config from istiod to generate a integer table, the value of table is endpoints index;
+
+
+<li>Base l4 or l7 info compute a hash value</li>
+use hash value get an table index, which is <code>index=hash%len(table)</code>;
+<li>Use index to access table</li>
+<code>ep=tables[index]</code>
+</ol>
+
+
+##### maglev lookup table
+Use a compact table, table entry is bit width of endpoint count.
+For the table size default is 16381.
```

Comment:
Is there a reference here?

Reply:
I actually refer to cilium and envoy project for the size of this value.